Add support for named exports

hedgepigdaniel opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

## Summary

See https://github.com/gregberge/loadable-components/issues/245

Currently to create a loadable component, the module that is asynchronously loaded must export the component as the default export (or be a commonjs module exporting the component as `module.exports`). This is problematic/limiting because:
- There are good reasons to avoid default exports (prevents grepping a codebase to find usages of an export, prevents import auto completion from working in most/all IDEs)
- It is not possible to make [multiple loadable components using the same split point](https://github.com/gregberge/loadable-components/issues/245#issuecomment-564467163) (instead one must hard code chunk names with webpack "magic comments" - not a standard feature)

## Detailed choices

See https://github.com/gregberge/loadable-components/issues/245 for preliminary discussion of the API
- `loadable` and `loadable.lib` still accept the import creator as the first argument. This means that for the common/simple case of `loadable(() => import('./xyz'))`, there is no breaking change, instead options are added.
- `resolveComponent` is optional and defaults to the existing behaviour. Again, no breaking change, but typescript type inference (for the props of the loadable component) does not work unless it is specified. I think this is a consequence of the CommonJS compatibility.

## Benefits

### Ability to use named exports (including on the server side)
Use `resolveComponent` to select the export

### Ability to create multiple loadable components from a single split point
```typescript
// split.tsx
import React from 'react';

type Props1 = {
  prop1: string;
};

export const Component1 ({ prop1 }: Props1): JSX.Element => (
  <span>{prop1}</span>
);

type Props2 = {
  prop2: string;
};

export const Component2 ({ prop2 }: Props2): JSX.Element => (
  <span>{prop2}</span>
);
```
```typescript
import loadable from '@loadable/component';

const LoadableComponent1 = loadable(() => import('./testComponent'), {
  resolveComponent: (mod) => mod.Component1,
});

const LoadableComponent2 = loadable(() => import('./testComponent'), {
  resolveComponent: (mod) => mod.Component2,
});
```

####  `resloveComponent` has access to props
```typescript
import loadable from '@loadable/component';
const DynamicLoadableComponent = loadable(() => import('./xyz'), {
  resolveComponent: (mod, props) => mod[props.export],
});
```

### Typescript type inference
When using `resolveComponent`, typescript is able to infer the type of the Props of the loadable component (previously they were always `unknown`). I think the default type/implementation of `resolveComponent` breaks type inference somehow by supporting commonjs default exports.

#### Before
```typescript
// testComponent.tsx
import React from 'react';

type TestProps = {
  prop1: string;
};

export default ({ prop1 }: TestProps): JSX.Element => (
  <span>{prop1}</span>
);
```
```typescript
// loadableTestComponent.ts
import loadable from '@loadable/component';

// Inferred type: `LoadableComponent<unknown>`  :(
const LoadableComponent = loadable(() => import('./testComponent'));
```

#### After
```typescript
// testComponent.tsx
import React from 'react';

type TestProps = {
  prop1: string;
};

export default ({ prop1 }: TestProps): JSX.Element => (
  <span>{prop1}</span>
);
```
```typescript
// loadableTestComponent.ts
import loadable from '@loadable/component';

// Inferred type: `LoadableComponent<TestProps>` :)
const LoadableComponent = loadable(() => import('./testComponent'), {
  resolveComponent: (mod) => mod.default,
});
```

## Test plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

Existing tests have been changed where necessary and new tests have been added for new functionality

hedgepigdaniel commented:
Amended types: https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...hedgepigdaniel:loadable-named-exports?expand=1

Artoria-0x04 commented:
Will this feature be in next release ?

theKashey commented:
- there should be some updates to the readme/site about the new feature
- `watchdog` sounds a bit strange for me. So it is a secondary promise resolver, which might do nothing more, but delay loadable resolution a bit, but only in "async" mode. It sounds more like a `condition` or `guard`. For example, I am using something like it to "delay" lazy component resolution till extra `.css` files, previously "critically inlined" got loaded.

hedgepigdaniel commented:
> * there should be some updates to the readme/site about the new feature

I've added documentation to the website to go with each commit

> * `watchdog` sounds a bit strange for me. So it is a secondary promise resolver, which might do nothing more, but delay loadable resolution a bit, but only in "async" mode. It sounds more like a `condition` or `guard`. For example, I am using something like it to "delay" lazy component resolution till extra `.css` files, previously "critically inlined" got loaded.

I thought of "watchdog" because its like a parallel process that supervises the loading process and may intervene if it goes wrong. "guard" also sounds fine to me, so I've updated it to be called "guard". That probably fits better with delays (the guard took a while to let the promise through) vs timeouts (the guard gave up waiting for the promise and rendered the error state).

hedgepigdaniel commented:
Also, I've put the commit that changes the Babel plugin last.

So if you wanted to delay the breaking change, this could be released in two steps:
1. New features `resolveComponent` and `guard`, updates to documentation to stop recommending wrapping the import promise and instead recommend `guard` for delays and timeouts
1. Breaking change: the Babel plugin makes an error if the `import()` is wrapped

theKashey commented:
You have got the approval from me. I like the change to the API, and I like the implementation. @gregberge - please take a look.

hedgepigdaniel commented on packages/babel-plugin/src/index.js:
```diff
@@ -72,10 +115,17 @@ const loadablePlugin = api => {
-    if (callPaths.length === 0) return
+      if (!isFunctionAndOnlyReturnsImport(importCreator)) {
+      throw new Error(
+        'The first argument to `loadable()` must be a function with a single statement that returns a call to `import()`'
```
I added a link to the notes for `loadFn`, which in turn link to the docs for guard, delay/timeout, and resolveComponent

hedgepigdaniel commented on packages/component/src/createLoadable.js:
```diff
@@ -33,6 +34,23 @@ function createLoadable({ resolve = identity, render, onLoad }) {
     }
 
+    function resolve(module, props, Loadable) {
+      const Component = options.resolveComponent
```
Done. That prevents you from using `resolveComponent` with `loadable.lib`, but then there is no point doing that (since you get the entire module anyway).

hedgepigdaniel replied:
Also I had to bump up the maximum bundle size

gregberge commented on website/src/pages/docs/api-loadable-component.mdx:
```diff
@@ -10,20 +10,76 @@ order: 10
+
+**Note:** It is highly recommended that you do not do anything in this function except returning an `import()` statement.
+This is because such functionality [will not work](/docs/babel-plugin/#detection-of-wrapped-imports) when using the [Babel plugin](/docs/babel-plugin/) or during [server side rendering](/docs/server-side-rendering/).
+For most use cases you can instead use the [`guard`](http://localhost:8001/docs/api-loadable-component/#optionsguard) option (e.g. for delays/timeouts), or the [`resolveComponent`](http://localhost:8001/docs/api-loadable-component/#optionsresolvecomponent) option (e.g. for wrapping the imported component).
```
Wrong link

hedgepigdaniel replied:
fixed

gregberge commented:
Hello guys! @hedgepigdaniel @theKashey 

First, thanks for your work on this feature! It is great!

I think we could make the API more generic and more powerful. I explain my idea:

```js
import loadable from '@loadable/component'

const Loadable = loadable(() => import('./components'))

// this function has to be synchronous, it is called on client and server
Loadable.resolveProps = ({ $component, $module, props, ctx }) => {
  // we got some data from context
  const data = ctx.books[props.id]
  // we are able to change the component
  return { $component: $module.Foo, props: { ...props, data } }
}

// this function will only be called on client-side, it can be asynchronous
Loadable.resolveContext = async ({ props }) => {
  // we actually load data dynamically with the component
  const book = await getBook(props.id)
  // we could also add a default delay
  await delay(200)
  // we are able to change the component
  return { books: { [props.id]: book  }
}

<Loadable id="10" />
```

On the server, we will be able to pass a context:

```js
const  book = await getBook(req.params.id)
const ctx = { books: { [req.params.id]: book }  }
const extractor = new ChunkExtractor({ statsFile, ctx })
```

If Suspense is available one day, resolveContext could be called on server-side. For now it is not possible.

What do you think?

gregberge commented:
Another proposal:

```js
import loadable from '@loadable/component'

const Loadable = loadable(() => import('./components'))

Loadable.resolveProps = () => {
  
}

// this function will only be called on client-side, it can be asynchronous
Loadable.getContext = async ({ props }) => {
  // we actually load data dynamically with the component
  const book = await getBook(props.id)
  // we could also add a default delay
  await delay(200)
  // we are able to change the component
  return { books: { [props.id]: book  }
}

// this function has to be synchronous, it is called on client and server
Loadable.resolve = ({ $component, $module, props, ctx }) => {
  // we got some data from context
  const data = ctx.books[props.id]
  // we are able to change the component
  return { $component: $module.Foo, props: { ...props, data } }
}

<Loadable id="10" />
```

On the server, we will be able to pass a context:

```js
const  book = await getBook(req.params.id)
const ctx = { books: { [req.params.id]: book }  }
const extractor = new ChunkExtractor({ statsFile, ctx })
```

If Suspense is available one day, resolveContext could be called on server-side. For now it is not possible.

What do you think?

gregberge commented:
Oups I closed it!

theKashey commented:
- Data loading is great, but it should be like a "single point of truth". Ie `getContext` should be also called - on the clientside, as well as on SSR.
- Data loading is important, and it could work "this" way after Suspense, but right now it has to be finished before the render, and bound to the routing.
- `getContext` might use some hooks(`useContext`) to provide data for a component, as well as `suspense` features in the future, but in __sync__ way, so it should be more compatible with __today__.

hedgepigdaniel commented:
hey, thanks for taking a look at this.

> What do you think?

- `getContext` is first defined on a particular loadable component, but on the server side it's passed to the entire react tree? This seems imconsistent.
 - Is `Loadable.resolve` the part that can resolve named exports? If so, how is it possible that it receives the resolved component as a prop already?
 - having options assigned imperatively (e.g. `Loadable.resolve = ...`) will make it hard for type inference to know the type of the loadable component

Having a way of passing extra props to the component seems neat, but I'm not sure how to do it consistently without suspense. If you have to set up 2 different methods for SSR and client side, I suspect using an external state management instead (e.g. Redux + route based data loading) would work better.

gregberge commented:
You are right. Data fetching is another problem, not our business here.

Your implementation is nice. I am wondering if guard is enough explicit or not.

theKashey commented:
Well, let's think what it is:
- it is a `guard`(as seen in `type guards`, or more like `xstate` guards, which prevent state transition)  - it might stop something
- it is a `check` - it might perform some operation, or wait for some operation before resolving import promise
- it a timeout - it might reject, causing component to fall into the error state
- it is __only__ for clientside, only for __async__, and it also could be async
- and we don't care about the result, only about "open" or "close" mode.

It could be a `guard`, it could be a `gateway`.

Another question - why we might use it:
- to emulate `time-in`, like "flash of loading content". The better option would be to use `useTransition`, which is not yet "official", but roughly reflect how one should handle a situation like this.
- to emulate `time-out`, but it could be done in another, more suspense friendly, way. Like `Suspense fallback` might change some state after timeout, causing Application to display an error state.
- to wait for some "external" condition. I personally use it with critical CSS extraction - like before "resolving" split code I have to "reinject" slit css. A quite rare case.

---

So what about just not mark this option as an experimental? I am pretty sure not many people were using this functionality so far, as long as they would had their SSR broken, and `React.lazy` does not provide anything like this, and people are "ok" with it...

hedgepigdaniel commented:
Another option is to have options more specifically tailored to common use cases. For example:
 - Add built in `minDelay` and `timeout` options (probably 95% of the use cases for `guard`)
 - rename `guard` to something related to loading extra modules/requests... e.g. `loadExtra` or `sideload` (probably 95% of the remaining use cases)

Marking it as experimental or simply not adding the option is also fine (and therefore not making the babel plugin complain basically the first 2 commits only). That still makes named exports possible, which is the most important thing. 

My thinking in including that change here is that it is confusing that the current `loadFn` works differently on the server side - delaying the promise works as you would hope, but changing its return value fails spectacularly. That is a trap for users, and trying to use named exports with the current API is one way you can fall into that trap, as in https://github.com/gregberge/loadable-components/issues/245.

theKashey commented:
> rename guard to something related to loading extra modules/requests

Which would not be called if this components was SSRed and loaded in a "sync" way, so please don't put any valuable logic inside. 😒

From a customer point of view it would be great to have just a global option to control `minDelay`/`timeout`, both of them could be "fixed" from the outside of Loadable component. In a bit harder way, but could.

hedgepigdaniel commented:
OK, I've separated out the `guard` part into https://github.com/gregberge/loadable-components/pull/511, since that seems like its controversial/unclear and its not the most important thing.

So now this is just adding the `resolveComponent` option, which is the minimum for named exports to work.

hedgepigdaniel commented:
bump @gregberge 

What do you think about merging this as is (i.e. no breaking change, just the new `resolveComponent` option)?

gregberge commented:
@hedgepigdaniel yes I think we can merge the resolveComponent part at first.

theKashey commented:
Back online. Where are we?

gregberge commented:
@theKashey I will try to merge it this week-end

gregberge commented:
#97 shows us how it is not natural for users to use `resolveComponent` instead of doing it in the function. 

I have another idea. Why not supporting only async function, we could replace the `async` function to a synchronous one and replace all `await import()` by a synchronous one. What do you think?

theKashey commented:
But it will be asynchronous in the "client-side" space, so we still have to support sync and async.

gregberge commented:
> But it will be asynchronous in the "client-side" space, so we still have to support sync and async.

I mean:

```js
const Loadable = loadable(async () => {
  const { Foo } = await import('x')
  return Foo
})
```

Will be transformed in:

```js
const Loadable = loadable(() => {
  const { Foo } = syncImport('x')
  return Foo
})
```

What do you think?

gregberge commented:
So I will explain my vision:

`resolveComponent` is not intuitive for users and make the library more complicated than it should. The usage is not natural.

The natural usage is to use that syntax:

```js
const Loadable = loadable(async () => {
  const { Foo } = await import('x')
  return Foo
})
```

With this syntax, we can easily detect that the user extract `{ Foo }` from `import('x')`. It means that we have the information.

The idea is to infer the synchronous operation:

```js
function load() {
  const { Foo } = importSync('x')
  return Foo
}
```

### Complex cases

If we allow an async function, users would be able to do something much more complex:

```js
const Loadable = loadable(async () => {
  const { Foo } = await timeout(import('x'), 2000)
  return Foo
})
```

What do we have here, we can easily detect that there is a pattern `await <not an import>`. In this case we can't infer the synchronous operation, the user have to write it:

```js
const resolveComponent = ({ Foo }) => Foo
const Loadable = loadable(async () => {
  return resolveComponent(await timeout(import('x'), 2000))
}, { 
  loadSync(props, { importSync }) {
    return resolveComponent(importSync('x'))
  }
})
```

It has several advantages:

- In most of cases it does not require any code to work
- It supports multiple `import()` in a single function
- It supports custom components creation

### Custom component creation

```
const Loadable = loadable(async () => {
  const { data } = await import('./data')
  const { Hello } = await import('./Welcome')
  return () => <div><Hello /></div>
})
```

This custom component could work without any problem!

What do you think?

theKashey commented:
Ok, so here are my thoughts: 
- I am not sure I personally capable not only to create, but even to design, a babel plugin to do the job. That would be a very complex transformation, and we don't need it.
- But you can use `babel`/`ts` transform to rewrite `import` to `loadableRequire` in all places. 

That is all we need

- `loadableRequire` would detect the origin of a call (`LoadableComponent` will set a scope)
- if called from the user code - it shall behave like a normal `import`
- if it called from the `loadable` - it shall try to be synchronous, 
- if that is not possible - it shall 😄 throw a `Promise`.
- if it is possible - it shall report used module/chunk/mark to the caller (`loadable`)

gregberge commented:
@theKashey I don't get it. If the function is `async` then we can't do anything to make it synchronous except a babel transformation isnt'it? 

About the Babel transformation, I think I could write it.

theKashey commented:
> @theKashey I don't get it. If the function is async then we can't do anything to make it synchronous except a babel transformation isnt'it?

You are transforming __only__ `import`, not a "function".

However, 😢, in this case stuff like TS will stop you from doing that.

hedgepigdaniel commented:
> The natural usage is to use that syntax:
> 
> ```js
> const Loadable = loadable(async () => {
>   const { Foo } = await import('x')
>   return Foo
> })
> ```

With suspense this would be the clear path forward, but given that we still need to synchronously resolve the component on the server side, It can't work 100% like that (somewhere there has to be a compromise).

The main difference with this proposed API and what I've written is that it doesn't require a second function (`resolveComponent` or `loadSync`) to handle a case where `loadFn` simply returns a key in the module. Let's call this API "exports" (because its automatic for exports of the module), and the code as I've written it "default" (because its automatic for the default export).

- :heavy_check_mark: Automatic from only `loadFn`
- :large_orange_diamond: Manual with `resolveComponent`/`loadSync`
- :x: Not possible
-----
| | status quo | "default" | "keys"
|--|--|--|--|
| default export | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| directly* return named export | :x: | :large_orange_diamond: | :heavy_check_mark: |
| return named export in non obvious way | :x: | :large_orange_diamond: | :large_orange_diamond:  |
| also do async things on client side | :x: | :large_orange_diamond: | :large_orange_diamond:  |
| use helper function to return export | :x: | :large_orange_diamond: | :large_orange_diamond:  |


You could spend a very long time making the babel plugin more and more complex so that it can handle more of these cases automatically with only `loadFn`, but it would never work automatically 100% of the time. There will always be more ways to return a component in `loadFn` that the Babel plugin can't handle.

> `resolveComponent` is not intuitive for users and make the library more complicated than it should. The usage is not natural.

What is not natural about it?

I think its still an improvement on the current API. Currently `loadFn` can do async things, but these silently fail on the server side. This is a big problem. It is inconsistent and confusing for users, and causes bugs. Specifying `resolveComponent` is slightly annoying (2 functions instead of just one), but is 100% consistent with no traps (it always works the same, client side or server side). And it lets you use any component you can get synchronously from the imported module, which is currently not possible.

hedgepigdaniel commented:
> It has several advantages:

I think all these things are supported well enough with the `resolveComponent` API

>     * In most of cases it does not require any code to work

`resolveComponent` is not needed for when you use the default export, which is already most of the time (currently 100% of the time, because non default exports don't work). True this alternative is more automatic, but its also more complex and the boundary between what is automatic and what isn't is harder to define.

>     * It supports custom components creation
>     * It supports multiple `import()` in a single function

You can do this with `resolveComponent` as follows:
```javascript
// async.js

export { Apple } from './apple';
export { Banana } from './banana';
```

```javascript
// loadable.js

const Loadable = loadable(() => import('./async'), {
  resolveComponent: ({ Apple, Orange }) => () => <div><Apple /><Orange /></div>,
});
```

It's better to have only one `import()` since then there is only one split point, and bundlers like webpack know to create only one chunk rather than two.

theKashey commented:
Meanwhile in the parallel universe: Suspense would work for SSR as well. Eventually, and probably quite soon.

We should design a future API with that "future" in mind. This also includes stuff like non-components to be trackable throughout the library, and that's not only about `loadable.lib`

kaysonwu commented:
> So I will explain my vision:
> 
> `resolveComponent` is not intuitive for users and make the library more complicated than it should. The usage is not natural.
> 
> The natural usage is to use that syntax:
> 
> ```js
> const Loadable = loadable(async () => {
>   const { Foo } = await import('x')
>   return Foo
> })
> ```
> 
> With this syntax, we can easily detect that the user extract `{ Foo }` from `import('x')`. It means that we have the information.
> 
> The idea is to infer the synchronous operation:
> 
> ```js
> function load() {
>   const { Foo } = importSync('x')
>   return Foo
> }
> ```
> 
> ### Complex cases
> If we allow an async function, users would be able to do something much more complex:
> 
> ```js
> const Loadable = loadable(async () => {
>   const { Foo } = await timeout(import('x'), 2000)
>   return Foo
> })
> ```
> 
> What do we have here, we can easily detect that there is a pattern `await <not an import>`. In this case we can't infer the synchronous operation, the user have to write it:
> 
> ```js
> const resolveComponent = ({ Foo }) => Foo
> const Loadable = loadable(async () => {
>   return resolveComponent(await timeout(import('x'), 2000))
> }, { 
>   loadSync(props, { importSync }) {
>     return resolveComponent(importSync('x'))
>   }
> })
> ```
> 
> It has several advantages:
> 
> * In most of cases it does not require any code to work
> * It supports multiple `import()` in a single function
> * It supports custom components creation
> 
> ### Custom component creation
> ```js
> const Loadable = loadable(async () => {
>   const { name } = await import('./data')
>   const { Hello } = await import('./Welcome')
>   return () => <div><Hello /> {name}</div>
> })
> ```
> 
> This custom component could work without any problem!
> 
> What do you think?

It's great to support multiple import ideas.

gregberge commented:
@hedgepigdaniel the problem with your implementation is that it restricts a lot possibilities if you don't use the SSR.

Actually it is very easy, you can use anything in your async function. I want to keep that and to provide a way for user to fix their code for SSR. Adding a `loadSync` method is an easy migration path for the SSR user. And the non-SSR user keep all power.

We can't make an API more restrictive than `React.lazy`. The Babel magic is optional.

I am sorry because you worked a lot on this but it is very important and we have to switch to `loadSync` approach for the future of this library.

hedgepigdaniel commented:
> @hedgepigdaniel the problem with your implementation is that it restricts a lot possibilities if you don't use the SSR.
> 
> Actually it is very easy, you can use anything in your async function. I want to keep that and to provide a way for user to fix their code for SSR. Adding a `loadSync` method is an easy migration path for the SSR user. And the non-SSR user keep all power.

I gathered that - that's why I've separated this PR from https://github.com/gregberge/loadable-components/pull/511. With the change in this PR now, I don't think there are any restrictions added that were not there before.

In particular, you can still do anything you want in the importFn, and it still works on the client side - with this change as it currently stands, users do not need to migrate to anything, their existing code will work.

I'm not sure why `loadSync` is a better API than `resolveComponent` - if users want to start using named exports then they are going to have to change something in their code. With `resolveComponent` they only have to write the import once, and the only difference is that the extraction of the export from the module needs to be in a different function (`resolveComponent`). With `loadSync`, the user needs to import the same file twice (one async, one sync), which seems unecessary/confusing.

> We can't make an API more restrictive than `React.lazy`. The Babel magic is optional.

Again, I don't think there are any breaking changes or new restrictions in this PR as it is now. I personally would recommend doing some. e.g. with https://github.com/gregberge/loadable-components/pull/511, but that is not necessary for this PR and to have functioning named exports.

gregberge commented:
@hedgepigdaniel I don't agree. The `resolveComponent` implies to return only an `import` from the function. Else it does not make sense sense. Except if you only apply it for SSR, in this case it should be `resolveSyncComponent` and make things more complicated than it should.

hedgepigdaniel commented:
Not sure what you mean by "imply", but
- having `resolveComponent` in the API does not prevent users from doing anything they want in loadFn - they still can and it still works as much as it did before
- It probably does give a hint that returning something other than the import in `loadFn` is not a good idea. I think that's a good thing, because returning different things in `loadFn` does not work properly and cannot be fixed!

So yes, `resolveComponent` is intended to replace the practice of returning different things from `loadFn`. It's a necessary replacement because the current API doesn't work for SSR, and its backwards compatible because everything that currently works will keep working (i.e. the client side).

If your concern is not that there is a breaking change, then what is your concern? I don't think I understand. Do you want so solve the problem in a different way? Do you want to solve the problem at all?

gregberge commented:
I just read again the documentation of `React.lazy` and I changed my mind. It it explicitly said that it has to return an `import` with a `default` component.

Maybe we should keep this restriction. They also say that [named exports are not supported](https://reactjs.org/docs/code-splitting.html#named-exports).

That's said, my point is :

1. We make a change to only support a function that returns an `import`. We warn user that returning anything except an import function will not be supported in next major version.
2. We introduce the `resolveComponent` option to provide a solution for named exports
3. We implement [delay](https://loadable-components.com/docs/delay/), `fallbackDelayMs` is a good time.
4. We document how to set [timeout](https://loadable-components.com/docs/timeout/) directly in [webpack](https://webpack.js.org/configuration/output/#outputchunkloadtimeout)

Sorry for this discussion. After a lot of thinking, I think your API is the good one. Now let's implement things and add some warning to make the transition easier to v6.

Actually I have time (we are confined in France). I will try to work on it this week-end!

theKashey commented:
There could also a good thing to document/explain why `default exports` are bad for common import/exports, but good for code splitting. Well - [Formats over Functions](https://dev.to/swyx/formats-over-functions-2em7), the same thing why it's expected to have _redux selectors_ in `selectors.js`.
Probably it would be also great to have a few examples where named exports could be good for code splitting as well, and chunk exporting _more that one thing_ is a good example. Or it's actually bad? Why you need other things, when you actually dont?

I reckon this less technical change, and more about "how it actually should work", and thus it's really important to get it right.

hedgepigdaniel commented:
Hooray, great to hear!

> That's said, my point is :
> 
>     1. We make a change to only support a function that returns an `import`. We warn user that returning anything except an import function will not be supported in next major version.
> 
>     2. We introduce the `resolveComponent` option to provide a solution for named exports
> 
>     3. We implement [delay](https://loadable-components.com/docs/delay/), `fallbackDelayMs` is a good time.
> 
>     4. We document how to set [timeout](https://loadable-components.com/docs/timeout/) directly in [webpack](https://webpack.js.org/configuration/output/#outputchunkloadtimeout)

Sounds like a good plan. Only issue is that 2 has to happen before 1, otherwise there will be no way to implement delay without generating warnings.

Re 4 - I didn't know it was possible to specify a timeout in `import()` - how do you do that?


How about we start with the `resolveComponent` functionality (i.e. merge this PR). Then task 1 will be unblocked, and we can think about how design a good API for delay and timeout before adding warnings/errors for users using the old API? https://github.com/gregberge/loadable-components/pull/511 would need a few changes to fit into this plan - perhaps we can discuss that there once this is merged?

> Probably it would be also great to have a few examples where named exports could be good for code splitting as well, and chunk exporting more that one thing is a good example. Or it's actually bad? Why you need other things, when you actually dont?

Hard to answer without looking at the use case I think - I guess its useful if as the developer you know that there a a few components (small unique code size for each one) that users usually use all/many of them, or none at all. That way it might be a good idea to make the first chunk slightly bigger and eliminate the delay for the other components. I did add an example in this PR (something to do with apples and oranges - but at least it illustrates the idea of having one split point with multiple components. Do you think that's enough, or it should be something else?

gregberge commented:
> Re 4 - I didn't know it was possible to specify a timeout in `import()` - how do you do that?

https://webpack.js.org/configuration/output/#outputchunkloadtimeout

> Hard to answer without looking at the use case I think - I guess its useful if as the developer you know that there a a few components (small unique code size for each one) that users usually use all/many of them, or none at all. That way it might be a good idea to make the first chunk slightly bigger and eliminate the delay for the other components. I did add an example in this PR (something to do with apples and oranges - but at least it illustrates the idea of having one split point with multiple components. Do you think that's enough, or it should be something else?

Naming imports are not a good idea because you don't have Tree Shaking. We have to warn about it on the documentation.

As soon as I have time, I will merge this one and work on it. Last week-end I was working on SVGR. One project per week-end :).

theKashey commented:
> Naming imports are not a good idea because you don't have Tree Shaking.

That's not quite bound to this case. It's more about - you will not _get_ treeshaking if you have more that one thingy exported from a module, and more that one thingy consumed.
But, if it's always one, and always use _through_ the loadable - why not to use default?

gregberge commented:
> But, if it's always one, and always use _through_ the loadable - why not to use default?

Yeah, the point is maybe we just have to "not support named exports". Creating a module wrapper is not a big deal and it is a good practice for tree shaking.

hedgepigdaniel commented:
Ok, so it sounds like we are trying to work out what is the use case for named exports. Here's a few use cases (using a module wrapper with a default export breaks these use cases):

### 1. Project where named exports are used for DX reasons

Default exports encourage the same thing to have different names in different files. This is confusing, prevents easy refactoring, prevents good support by tooling (e.g. import auto complete), and adds to cognitive load. See this [short and sweet post](https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad) for more info.

### 2. Connecting multiple loadable components to the same split point

Take an example that there is a modal that the user may open, and the modal that has tabs inside it. Users usually explore all the tabs after they open the modal. There are 5 tabs. 2 tabs are visible to all users, but 3 tabs are only accessible to administrators.. The content of the tabs is reasonably small (not much data to download), but each request still adds latency.

In this case, it might be a good optimization to use the same split point for all of the 3 components used in the administrator tabs. That way if a user opens one of the tabs, the code for all the tabs is downloaded, so there is no extra latency when they switch to other tabs.

It's not possible for webpack to determine which split points will be loaded at the same time at runtime, so allowing multiple named exports in a single points allows the user to give webpack that information so that it can make a more optimal choice of which code to put in which chunks.

There seems to be a few users interested in this ability: https://github.com/gregberge/loadable-components/issues/245#issuecomment-564467163

---

To put it really simply, javascript supports named exports, webpack supports named exports, so loadable-components should support named exports. The fact that it currently doesn't is surprising and inconsistent with the rest of the ecosystem.

Re tree shaking, it seems like its a fundamental limitation with the current javascript that `import()` does not support statically importing a particular export - it always returns the entire module. I think this problem is out of scope for this PR (and this library). The user is free to use 1 default export, 1 named export, or multiple named exports. The user writes the code that does `import('./xyz')`, so they are responsible for what that does. This problem already exists anyway - even though the user must use a default export with loadable-components, they can still add more (unused) named exports to the imported module if they want, and tree shaking will not remove them. Perhaps one day there will be an `import async { Thing } from './module';`.

theKashey commented:
Ok. So in short - the main advantage for named exports is the ability to use them "more often" without overthinking the problem and extra boilerplate.
Ie __just use__.

gregberge commented:
@hedgepigdaniel the chunk name resolves the second issue. Webpack group chunks with the same name.

Also the case you mentioned is not the nominal one, I think most of time users will use it for wrong reasons and just lose the tree shaking.

hedgepigdaniel commented:
No, chunk names does not resolve the issue. Chunk names lets you force webpack to put things in the same chunk, but:
- it prevents webpack from combining that with other chunks if it sees fit (as it does, to make the best compromise the number of requests vs the amount of data downloaded).
- That feature is a non standard webpack specific protocol, and it uses JSON inside comments which is awful.
- That means the user is responsible for choosing unique chunk names, which is insane

The idiomatic way to have multiple things behind a single split point is to export multiple things from one module, with named exports. No non standard features, no hacks to force a specific compiler to do what you want, just using the standard features of the language.

hedgepigdaniel commented:
You only "lose" tree shaking to the extent that you put *multiple* named exports in one file anyway - but loadable-components is currently broken even with a *single* named export. The use cases are already explained in the original PR description and the linked issue. This is clearly causing pain and confusion (and bugs in production), and there is a simple fix right here. @gregberge It's been almost 4 months now since this PR was opened, and you've already said twice in this thread that you will merge it as it is, but every 2 weeks you change your mind. How much longer do we have to continue this discussion?

gregberge commented:
@hedgepigdaniel since I am not very involved in the project it takes time and I am sorry for that. Yeah it is broken but we have two choices:

- Make it work
- Make it impossible

https://reactjs.org/docs/code-splitting.html#named-exports

"React.lazy currently only supports default exports." The currently makes me think that it is not definitive.

Let's make it possible. Also @hedgepigdaniel are you interested to be more involved in the project? Maintainer?

hedgepigdaniel commented:
This project? I thought you were the author of it?

Sure, I'm happy to help with maintaining. HMU if you want to chat about it.

gtrufitt commented:
Morning @gregberge @hedgepigdaniel thanks for your work on this - reading through the thread it seems like this has been ready to ship for a while. From a consumer perspective, it'd be great to see that happen as it's probably a blocker for us considering we'd have to refactor our project to use default exports.

theKashey commented:
Meanwhile...
Long story short but at the _"real work project"_ we are using __only__ names imports. And just a few days ago I've got a good example of exporting multiple entities from one chunk.
- there is a `Page`
- the `Page` defines data query _above_ it, so it could load js in a parallel with data
- the `Page` contains a subroute, which initialization code is a part of the `Page`
- but that code would be executed only after `Page` loads and that might take a while.

So - if you are landing to a subroute, you have to "see" how parent page is loading.

### The right fix for the problem
Refactor all routing, so the problem will not exists.

### The "faster" fix for the problem
Hoist `Route` a bit above, and "pick" one more export from the chunk. The case is solved, roughly 1 minute for all.

- In other words - named exports, as well as multiple imports from one file, are very refactoring friendly.
- In other other words - we should not bound our decisions to "React", as long as codesplitting is just a generic module management pattern. As long as we are not using the same module system as Facebook - their decisions might not be very applicable to our webpack world. 

👍 make it possible.

lkaratun commented:
Would love to see this merged soon! I really want to use the package in my org's codebase but we have a ton of named exports

smoothie99 commented:
Hi @hedgepigdaniel and @gregberge what is the plan for this?

zeppelinen commented:
I also would love to see it merged in the main tree.

theKashey commented:
Then let's do it!
(reading PR again)

clgeoio commented:
+1, would love to see this merged, happy to help fix any blockers

kyarunts commented:
+1, looking forward to seeing this merged as soon as possible :)

theKashey commented on packages/component/src/createLoadable.js:
```diff
@@ -33,6 +35,19 @@ function createLoadable({ resolve = identity, render, onLoad }) {
+      const Component = options.resolveComponent
+        ? options.resolveComponent(module, props)
+        : defaultResolveComponent(module)
+      if (options.resolveComponent && !ReactIs.isValidElementType(Component)) {
```
- [ ] to double check how this plays with `loadable.lib`

theKashey commented:
Ok. So the only missing piece is to update `definitely-typed`

kelly-tock commented:
where are the docs for this?  I don't see it on the site anywhere?

hedgepigdaniel commented:
https://loadable-components.com/docs/api-loadable-component/#optionsresolvecomponent

kelly-tock commented:
Ah yes thanks!
